File: bookshelf/services/sqlite_cache.py
__all__ = ["SQLiteCache"]

import json
import sqlite3
from datetime import date, timedelta
from pathlib import Path
from typing import Any

from bookshelf import get_cache_root


class SQLiteCache:
    def __init__(
        self,
        path: Path = None,
        expiry: int | None = 14,
    ):
        self.expiry = expiry
        self.con = sqlite3.connect(path or get_cache_root() / "cache.sqlite")
        self.con.row_factory = sqlite3.Row

        self.con.execute("CREATE TABLE IF NOT EXISTS queries (query, response, query_date);")
        self.delete()

    def select(self, query: str) -> dict[str, Any]:
        if self.expiry:
            expiry = date.today() - timedelta(days=self.expiry)
            cursor = self.con.execute(
                "SELECT * FROM queries WHERE query = ? and query_date > ?;",
                (query, expiry.isoformat()),
            )
        else:
            cursor = self.con.execute("SELECT * FROM queries WHERE query = ?;", (query,))
        results = cursor.fetchone()
        if results:
            return json.loads(results["response"])
        return {}

    def insert(self, query: str, response: dict[str, Any]) -> None:
        self.con.execute(
            "INSERT INTO queries (query, response, query_date) VALUES (?, ?, ?);",
            (query, json.dumps(response), date.today().isoformat()),
        )
        self.con.commit()

    def delete(self) -> None:
        if not self.expiry:
            return
        expiry = date.today() - timedelta(days=self.expiry)
        self.con.execute("DELETE FROM queries WHERE query_date < ?;", (expiry.isoformat(),))
        self.con.commit()

File: openlibrary/src/main/kotlin/github/buriedincode/openlibrary/SQLiteCache.kt
package github.buriedincode.openlibrary

import java.nio.file.Path
import java.sql.Date
import java.sql.DriverManager
import java.time.LocalDate
import kotlin.use

data class SQLiteCache(val path: Path, val expiry: Int? = null) {
    private val databaseUrl: String = "jdbc:sqlite:$path"

    init {
        this.createTable()
        this.cleanup()
    }

    private fun createTable() {
        val query = "CREATE TABLE IF NOT EXISTS queries (url, response, query_date);"
        DriverManager.getConnection(this.databaseUrl).use {
            it.createStatement().use {
                it.execute(query)
            }
        }
    }

    fun select(url: String): String? {
        val query = if (this.expiry == null) {
            "SELECT * FROM queries WHERE url = ?;"
        } else {
            "SELECT * FROM queries WHERE url = ? and query_date > ?;"
        }
        DriverManager.getConnection(this.databaseUrl).use {
            it.prepareStatement(query).use {
                it.setString(1, url)
                if (this.expiry != null) {
                    it.setDate(2, Date.valueOf(LocalDate.now().minusDays(this.expiry.toLong())))
                }
                it.executeQuery().use {
                    return it.getString("response")
                }
            }
        }
    }

    fun insert(url: String, response: String) {
        if (this.select(url = url) != null) {
            return
        }
        val query = "INSERT INTO queries (url, response, query_date) VALUES (?, ?, ?);"
        DriverManager.getConnection(this.databaseUrl).use {
            it.prepareStatement(query).use {
                it.setString(1, url)
                it.setString(2, response)
                it.setDate(3, Date.valueOf(LocalDate.now()))
                it.executeUpdate()
            }
        }
    }

    fun delete(url: String) {
        val query = "DELETE FROM queries WHERE url = ?;"
        DriverManager.getConnection(this.databaseUrl).use {
            it.prepareStatement(query).use {
                it.setString(1, url)
                it.executeUpdate()
            }
        }
    }

    fun cleanup() {
        if (this.expiry == null) {
            return
        }
        val query = "DELETE FROM queries WHERE query_date < ?;"
        val expiryDate = LocalDate.now().minusDays(this.expiry.toLong())
        DriverManager.getConnection(this.databaseUrl).use {
            it.prepareStatement(query).use {
                it.setDate(1, Date.valueOf(expiryDate))
                it.executeUpdate()
            }
        }
    }
}
